### ISSUE TITLE
porting the desugarer to the new typechecker

### ISSUE THREAD

**Issue by liamoc** [NICTA/cogent#40]
This is actually a substantial bit of work.

We need to:

- [ ] Add a pass to add "promote" nodes for subtyping (the existing "promote" infrastructure is no longer needed). I (@liamoc) can do this. I'm going to try and do this at the same time as core typechecking or just after.
- [ ] Adapting the core AST to include widen (for integer) and promote (for subtyping) nodes.
- [ ] Adapt the existing desugarer to deal with the new type system. I think we should make it so that instead of outputting `UntypedExpr t v VarName` it produces `UntypedExpr t v (VarName, RawType)`. That should work nicely and mean we can get rid of most of the actual type checking in the core type checker, which might be good.

Once this is done, I can write the core type checker.
